lib/puppet/type/agent_upgrade.rb

```diff
@@ -0,0 +1,30 @@
+Puppet::Type.newtype(:agent_upgrade) do
+  ensurable
+
+  newparam(:name, :namevar => true)
+
+  newparam(:source) do
+    newvalues /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+    validate do |value|
+      if value !~ /^((https?|puppet):\/\/)|\\\\\S+\\\w+|[a-z]:\\/i
+        fail "Please provide a valid http(s), puppet or unc path"
+      end
+      if value =~ /^https?/ && value !~ URI::regexp
+        fail("Please provide a valid URI")
```

Comment:
The agnostic sense is that I want to upgrade all of my clients to the latest. Most folks download once and make it available there. That means now that I need to fully qualify source to the file that I need a selector/case statement for all platforms to be sure I get the right file for each platform. 


cc @joshcooper @MikaelSmith @kylog